### ISSUE TITLE
Vertical tabs

### ISSUE THREAD

**Issue by kekscode** [Jelmerro/Vieb#113]
**Checklist**
- [x] I searched the `:help` documentation using `/` for helpful information
- [x] I have read the [frequently asked questions](https://github.com/Jelmerro/Vieb/blob/master/FAQ.md)
- [x] I did not find any similar issues in the [issue tracker](https://github.com/Jelmerro/Vieb/issues)

**Addition or change**
I come from qutebrowser and i like vieb a lot so far, because it is much faster and not laggy at all. What i am missing the most right now is the possibility to have vertically stacked tabs on the left of the browser window. The reason for this is:

* Browsers on modern widescreen monitors usually leave a lot of empty screen estate left and right from a webpage's content, because wesites are usually designed to be used in the vertical direction
* Having a lot of open tabs with long titles is not a problem if they're vertically aligned. 50+ open tabs are not a problem.

**Alternatives considered**
I don't have a real alternative to offer. I am really used to vertically stacked tabs and it is the primary feature i am looking for when choosing a browser.

**Comment by Jelmerro** [comment]
I think this can already be done with custom themes, though someone has yet to make one. All built-in colorschemes can be found [here](https://github.com/Jelmerro/Vieb/tree/master/app/colors). The default one is actually all the CSS of Vieb, so you should be able to get a good overview of the styling from that. The flipped theme is an example of moving the GUI elements to the bottom of the window, I have no doubts that something similar can be done for moving the tabs to the left. I'll do some testing of this myself, but I expect users to be able to make these themes according to their own requirements. And if you feel that more people would like your theme, you can request it to be added with a PR.

**Comment by Jelmerro** [comment]
I pushed a new commit with the initial version of the "verticaltabs" colorscheme. If you follow the instructions from the readme to run Vieb from source you can test it out. Feel free to make changes to the CSS, I'm not the end user of this feature after all. Here's how it currently looks:
![](https://user-images.githubusercontent.com/1696674/102724678-82080200-4311-11eb-9d16-d899ccb48da9.png)

**Comment by kekscode** [comment]
Amazing! Thank you a lot for the quick response and proof of concept. I will give it a try as soon as possible and provide proper feedback.

**Comment by kekscode** [comment]
Strange: I updated to Mac OS BigSur today and now i wanted to test the CSS modification, but vieb stopped working on BigSur. I tried the latest release (still installed and worked yesterday on Mac OS Catalina just fine) and i did the build following the instructions in the README).

The following is happening:
Electron starts, the menubar is visible, but no main window gets initialized. It is **not** important if i doubleclick the application icon or if i use `npm start` from inside the cloned repository, the result is the same. "Quit Vieb" is accessible in the menubar (the menubar is not part of the main window on Mac, that's why it is reachable), but i cannot quit it with this option - i have to force it by killing the process). The about vieb window can be triggered and opens perfectly, though. All other application-specific options in the menubar are either greyed out or seem to do nothing.

I don't see any errors in the console output (npm start), though. I'm a bit puzzled here and i lack know how in the electron/javascript domain. I see some open issues on the electron repository regarding big sur and apple silicon, but these shouldn't affect my machine which is running on Intel. Otherwise no known issues apparently, especially not describing what i see on my machine.

Before i try to put debug print statemets to the index.js, is there a good way on how to debug this properly? Maybe some debug/verbose switch or some documentation on how to debug an electron app for newbies like me?

**Comment by Jelmerro** [comment]
You can start with `--debug` if you haven't done so already.

**Comment by kekscode** [comment]
At least this gives me a devtools window and a "broken" vieb window. Broken means: I can switch modes, but nothing else seems to work (no ex commands, no search/navigate in explore, ...) and the debug console in the devtools window throws various `Uncaught (in promise) TypeError: Cannot read property [...]` errors of various sorts depending on what i try to do.

At least i should be able to tinker with it now. No errors or anything else on stdout/stderr though. Just:

```
❯ npm start -- --debug

> vieb@3.2.0-dev start
> electron app "--debug"
```

![vieb-dev-npm-start_debug](https://user-images.githubusercontent.com/803252/102834710-1eef9b80-43f5-11eb-85dd-38d4a862e143.png)

**Comment by Jelmerro** [comment]
I can't reproduce this anywhere, but it seems like it can't find the current tab. This could be due to many things, but given that it started after the update, it feels related to Big Sur in some way. Unless you find a way to reproduce this reliably on other operating systems, I'm afraid you're on your own with solving this, as I don't own a mac.

**Comment by kekscode** [comment]
Yes, i'll try to figure that out. Too bad this issue went off-topic now. Do you want to cleanup this issue somehow like opening a new one for the big sur issue and delete the big sur related posting from this issue? (i don't know if this is even possible here in github, but it annoys me a bit that things are mixed up now in your projects issue tracker)

**Comment by Jelmerro** [comment]
I don't really mind, it's a relevant discussion because it prevents you from testing this new feature. Though I would recommend to open a fresh issue for this (and do so in the future as well). But there's no need to delete comments.

**Comment by kekscode** [comment]
Back on topic on the vertical tabs colorscheme and my initial feedback:

- The feature looks great and saves me A LOT of screen estate
- Usability is also great (Navigation with J/K comes natural for me and tab cycling - if turned on - works flawless)
- The scrollbar is useful as a visual indicator if the list grows larger than the vertical screen estate allows
- Shrinking and expanding the window on the horizontal direction is working well and the relative shrinking/expanding of the tabs according to the window width seems well balanced
- `mintabwidth` and `taboverflow` does not seem to do anything, which would be logical, because having vertically stacked tabs does not create the issues these two options deal with. So i guess this is a good thing, because switching back to the normal horizontal tabs will automatically use these two options and switching to vertically stacked tabs just works without any side effects - great :)

Not sure if all gui...tabbar options work properly, though. Did you already test these options or should i examine that a bit further? (I never use fullscreen myself, that's why i didn't test this yet)

**Comment by Jelmerro** [comment]
I have tested fullscreen (both user activated and video) briefly and it worked as expected. We can of course keep improving the layout if we later discover any other issues.

I can't reproduce the issue you mention with regards to the viebrc not working for colorschemes. The viebrc I tested with simply contains this: `colorscheme verticaltabs` Which seems to activate the verticaltabs layout reliably on startup.
